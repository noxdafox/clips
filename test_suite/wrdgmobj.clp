
;;;======================================================
;;;   Number Puzzle Problem
;;;     
;;;     Solves the number puzzle problem in which
;;;
;;;          GERALD 
;;;        + DONALD
;;;          ------
;;;        = ROBERT
;;;
;;;     CLIPS Version 6.0 Object Example
;;;
;;;     To execute, merely load, reset and run.
;;;     This example takes alot of memory to execute.
;;;======================================================

(defclass combination (is-a USER)
   (role concrete)
   (pattern-match reactive)
   (slot number (create-accessor write))
   (slot letter (create-accessor write)))

(defclass number (is-a USER)
   (role concrete)
   (pattern-match reactive)
   (slot value (create-accessor write)))

(defclass letter (is-a USER)
   (role concrete)
   (pattern-match reactive)
   (slot value (create-accessor write)))

(defrule startup
  =>
  (printout t crlf "The problem is" crlf crlf)
  (printout t "   GERALD" crlf)
  (printout t " + DONALD" crlf)
  (printout t "   ------" crlf)
  (printout t " = ROBERT" crlf crlf)
  (make-instance of number (value 0))
  (make-instance of number (value 1))
  (make-instance of number (value 2))
  (make-instance of number (value 3))
  (make-instance of number (value 4))
  (make-instance of number (value 5))
  (make-instance of number (value 6))
  (make-instance of number (value 7))
  (make-instance of number (value 8))
  (make-instance of number (value 9))
  (make-instance of letter (value G))
  (make-instance of letter (value E))
  (make-instance of letter (value R))
  (make-instance of letter (value A))
  (make-instance of letter (value L))
  (make-instance of letter (value D))
  (make-instance of letter (value O))
  (make-instance of letter (value N))
  (make-instance of letter (value B))
  (make-instance of letter (value T)))

(defrule generate-combinations
  (object (is-a number) (value ?x))
  (object (is-a letter) (value ?a))
  =>
  (make-instance of combination (letter ?a) (number ?x)))

(defrule find-solution
  (object (letter D) (number ?d))
  (object (letter T) (number ?t&~?d))
  (test (= (mod (+ ?d ?d) 10) ?t))
  (object (letter L) (number ?l&~?d&~?t))
  (object (letter R) (number ?r&~?d&~?t&~?l))
  (test (= (mod (+ ?d ?d
                   (* 10 ?l) (* 10 ?l))
                100)
           (+ (* 10 ?r) ?t)))
  (object (letter A) (number ?a&~?d&~?t&~?l&~?r))
  (object (letter E) (number ?e&~?d&~?t&~?l&~?r&~?a))
  (test (= (mod (+ ?d ?d
                   (* 10 ?l) (* 10 ?l)
                   (* 100 ?a) (* 100 ?a))
                1000)
           (+ (* 100 ?e) (* 10 ?r) ?t)))
  (object (letter N) (number ?n&~?d&~?t&~?l&~?r&~?a&~?e))
  (object (letter B) (number ?b&~?d&~?t&~?l&~?r&~?a&~?e&~?n))
  (test (= (mod (+ ?d ?d
                   (* 10 ?l) (* 10 ?l)
                   (* 100 ?a) (* 100 ?a)
                   (* 1000 ?r) (* 1000 ?n))
                10000)
           (+ (* 1000 ?b) (* 100 ?e) (* 10 ?r) ?t)))
  (object (letter O) (number ?o&~?d&~?t&~?l&~?r&~?a&~?e&~?n&~?b))
  (object (letter G) (number ?g&~?d&~?t&~?l&~?r&~?a&~?e&~?n&~?b&~?o))
  (test (= (+ ?d ?d
              (* 10 ?l) (* 10 ?l)
              (* 100 ?a) (* 100 ?a)
              (* 1000 ?r) (* 1000 ?n)
              (* 10000 ?e) (* 10000 ?o)
              (* 100000 ?g) (* 100000 ?d))
           (+ (* 100000 ?r) (* 10000 ?o) (* 1000 ?b) (* 100 ?e) (* 10 ?r) ?t)))
  =>
  (printout t "A Solution is:" crlf crlf)
  (printout t "  G = " ?g crlf)
  (printout t "  E = " ?e crlf)
  (printout t "  R = " ?r crlf)
  (printout t "  A = " ?a crlf)
  (printout t "  L = " ?l crlf)
  (printout t "  D = " ?d crlf)
  (printout t "  O = " ?o crlf)
  (printout t "  N = " ?n crlf)
  (printout t "  B = " ?b crlf)
  (printout t "  T = " ?t crlf)
  (printout t crlf)
  (printout t "   " ?g ?e ?r ?a ?l ?d crlf)
  (printout t " + " ?d ?o ?n ?a ?l ?d crlf) 
  (printout t "   " "------" crlf)
  (printout t " = " ?r ?o ?b ?e ?r ?t crlf crlf))  
  
 
