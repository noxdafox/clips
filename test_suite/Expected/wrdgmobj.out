FIRE    1 startup: *

The problem is

   GERALD
 + DONALD
   ------
 = ROBERT

FIRE    2 generate-combinations: [gen1],[gen20]
FIRE    3 generate-combinations: [gen2],[gen20]
FIRE    4 generate-combinations: [gen3],[gen20]
FIRE    5 generate-combinations: [gen4],[gen20]
FIRE    6 generate-combinations: [gen5],[gen20]
FIRE    7 generate-combinations: [gen6],[gen20]
FIRE    8 generate-combinations: [gen7],[gen20]
FIRE    9 generate-combinations: [gen8],[gen20]
FIRE   10 generate-combinations: [gen9],[gen20]
FIRE   11 generate-combinations: [gen10],[gen20]
FIRE   12 generate-combinations: [gen1],[gen19]
FIRE   13 generate-combinations: [gen2],[gen19]
FIRE   14 generate-combinations: [gen3],[gen19]
FIRE   15 generate-combinations: [gen4],[gen19]
FIRE   16 generate-combinations: [gen5],[gen19]
FIRE   17 generate-combinations: [gen6],[gen19]
FIRE   18 generate-combinations: [gen7],[gen19]
FIRE   19 generate-combinations: [gen8],[gen19]
FIRE   20 generate-combinations: [gen9],[gen19]
FIRE   21 generate-combinations: [gen10],[gen19]
FIRE   22 generate-combinations: [gen1],[gen18]
FIRE   23 generate-combinations: [gen2],[gen18]
FIRE   24 generate-combinations: [gen3],[gen18]
FIRE   25 generate-combinations: [gen4],[gen18]
FIRE   26 generate-combinations: [gen5],[gen18]
FIRE   27 generate-combinations: [gen6],[gen18]
FIRE   28 generate-combinations: [gen7],[gen18]
FIRE   29 generate-combinations: [gen8],[gen18]
FIRE   30 generate-combinations: [gen9],[gen18]
FIRE   31 generate-combinations: [gen10],[gen18]
FIRE   32 generate-combinations: [gen1],[gen17]
FIRE   33 generate-combinations: [gen2],[gen17]
FIRE   34 generate-combinations: [gen3],[gen17]
FIRE   35 generate-combinations: [gen4],[gen17]
FIRE   36 generate-combinations: [gen5],[gen17]
FIRE   37 generate-combinations: [gen6],[gen17]
FIRE   38 generate-combinations: [gen7],[gen17]
FIRE   39 generate-combinations: [gen8],[gen17]
FIRE   40 generate-combinations: [gen9],[gen17]
FIRE   41 generate-combinations: [gen10],[gen17]
FIRE   42 generate-combinations: [gen1],[gen16]
FIRE   43 generate-combinations: [gen2],[gen16]
FIRE   44 generate-combinations: [gen3],[gen16]
FIRE   45 generate-combinations: [gen4],[gen16]
FIRE   46 generate-combinations: [gen5],[gen16]
FIRE   47 generate-combinations: [gen6],[gen16]
FIRE   48 generate-combinations: [gen7],[gen16]
FIRE   49 generate-combinations: [gen8],[gen16]
FIRE   50 generate-combinations: [gen9],[gen16]
FIRE   51 generate-combinations: [gen10],[gen16]
FIRE   52 generate-combinations: [gen1],[gen15]
FIRE   53 generate-combinations: [gen2],[gen15]
FIRE   54 generate-combinations: [gen3],[gen15]
FIRE   55 generate-combinations: [gen4],[gen15]
FIRE   56 generate-combinations: [gen5],[gen15]
FIRE   57 generate-combinations: [gen6],[gen15]
FIRE   58 generate-combinations: [gen7],[gen15]
FIRE   59 generate-combinations: [gen8],[gen15]
FIRE   60 generate-combinations: [gen9],[gen15]
FIRE   61 generate-combinations: [gen10],[gen15]
FIRE   62 generate-combinations: [gen1],[gen14]
FIRE   63 generate-combinations: [gen2],[gen14]
FIRE   64 generate-combinations: [gen3],[gen14]
FIRE   65 generate-combinations: [gen4],[gen14]
FIRE   66 generate-combinations: [gen5],[gen14]
FIRE   67 generate-combinations: [gen6],[gen14]
FIRE   68 generate-combinations: [gen7],[gen14]
FIRE   69 generate-combinations: [gen8],[gen14]
FIRE   70 generate-combinations: [gen9],[gen14]
FIRE   71 generate-combinations: [gen10],[gen14]
FIRE   72 generate-combinations: [gen1],[gen13]
FIRE   73 generate-combinations: [gen2],[gen13]
FIRE   74 generate-combinations: [gen3],[gen13]
FIRE   75 generate-combinations: [gen4],[gen13]
FIRE   76 generate-combinations: [gen5],[gen13]
FIRE   77 generate-combinations: [gen6],[gen13]
FIRE   78 generate-combinations: [gen7],[gen13]
FIRE   79 generate-combinations: [gen8],[gen13]
FIRE   80 generate-combinations: [gen9],[gen13]
FIRE   81 generate-combinations: [gen10],[gen13]
FIRE   82 generate-combinations: [gen1],[gen12]
FIRE   83 generate-combinations: [gen2],[gen12]
FIRE   84 generate-combinations: [gen3],[gen12]
FIRE   85 generate-combinations: [gen4],[gen12]
FIRE   86 generate-combinations: [gen5],[gen12]
FIRE   87 generate-combinations: [gen6],[gen12]
FIRE   88 generate-combinations: [gen7],[gen12]
FIRE   89 generate-combinations: [gen8],[gen12]
FIRE   90 generate-combinations: [gen9],[gen12]
FIRE   91 generate-combinations: [gen10],[gen12]
FIRE   92 generate-combinations: [gen1],[gen11]
FIRE   93 generate-combinations: [gen2],[gen11]
FIRE   94 find-solution: [gen66],[gen21],[gen79],[gen98],[gen85],[gen110],[gen47],[gen34],[gen53],[gen112]
A Solution is:

  G = 1
  E = 9
  R = 7
  A = 4
  L = 8
  D = 5
  O = 2
  N = 6
  B = 3
  T = 0

   197485
 + 526485
   ------
 = 723970

FIRE   95 generate-combinations: [gen3],[gen11]
FIRE   96 generate-combinations: [gen4],[gen11]
FIRE   97 generate-combinations: [gen5],[gen11]
FIRE   98 generate-combinations: [gen6],[gen11]
FIRE   99 generate-combinations: [gen7],[gen11]
FIRE  100 generate-combinations: [gen8],[gen11]
FIRE  101 generate-combinations: [gen9],[gen11]
FIRE  102 generate-combinations: [gen10],[gen11]
